numpyro/contrib/control_flow/cond.py

```diff
@@ -0,0 +1,80 @@
+from functools import partial
+
+from jax import device_put, lax
+
+from numpyro import handlers
+from numpyro.primitives import _PYRO_STACK, apply_stack
+from numpyro.contrib.control_flow.scan import PytreeTrace
+
+
+def _subs_wrapper(subs_map, site):
+    if isinstance(subs_map, dict) and site["name"] in subs_map:
+        return subs_map[site["name"]]
+    elif callable(subs_map):
+        rng_key = site["kwargs"].get("rng_key")
+        subs_map = (
+            handlers.seed(subs_map, rng_seed=rng_key)
+            if rng_key is not None
+            else subs_map
+        )
+        return subs_map(site)
+    return None
+
+
+def wrap_fn(fn, substitute_stack):
+    def wrapper(wrapped_operand):
+        rng_key, operand = wrapped_operand
+
+        with handlers.block():
+            seeded_fn = handlers.seed(fn, rng_key) if rng_key is not None else fn
+            for subs_type, subs_map in substitute_stack:
+                subs_fn = partial(_subs_wrapper, subs_map)
+                if subs_type == "condition":
+                    seeded_fn = handlers.condition(seeded_fn, condition_fn=subs_fn)
+                elif subs_type == "substitute":
+                    seeded_fn = handlers.substitute(seeded_fn, substitute_fn=subs_fn)
+
+            with handlers.trace() as trace:
+                value = seeded_fn(operand)
+
+        return value, PytreeTrace(trace)
```

Comment:
Returning only `value` here and not the `trace` eliminates the error, but I don't understand why yet... 😕

Reply:
This seems like a leak in PytreeTrace implementation, where we consider `rng_key` as an auxiliary (non traced) static data. Could you help me add the following logic at [this line](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/contrib/control_flow/scan.py#L39)?
```python
                        if key == "kwargs":
                            kwargs = site["kwargs"].copy()
                            if "rng_key" in kwargs:
                                kwargs["rng_key"] = None
                            aux_trace[name][key] = kwargs
                        else:
                            aux_trace[name][key] = site[key]
```
The reason that I didn't want to trace it is, under `scan`, those `rng_key`s will be collected, hence the output trace will store a batch of `rng_key`s, which does not make sense for a NumPyro `sample` site (which only requires 1 single key rather than a batch of keys). But I didn't anticipate that it leaks to some error like this. To be safe, the above code set `rng_key` to None in the collected trace. :)

Reply:
Thanks! This makes sense.